### PR TITLE
convert_model() slow conversion time fix

### PR DESCRIPTION
**Pull request by popovaan**
Root cause analysis: 
Non-default convert_model() parames are serialized to rt_info section. If "input_model" model is an object then it lead to serializing of the model object to IR. Example of such line:
<input_model value="&lt;tensorflow.python.saved_model.load.Loader._recreate_base_user_object.&lt;locals>._UserObject object at 0x7fcc14508f70>" />

For big models it leads to long conversion time.

Solution: Not to serialize input_model value to rt_info if input model is model object.

Ticket: 99306

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names
* [x]  Transformation preserves tensor names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests
* [x]  Transformation tests
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list
* [x]  Guide on how to convert the **public** model
* [x]  User guide update

### CODE REVIEW

File: tools/mo/unit_tests/mo/frontend_ngraph_test.py [hunk 1]
```diff
@@ -76,6 +76,19 @@ def test_main_error_log():
...
+def test_rt_info():
+    setup_env()
+    args = [sys.executable, '-m', 'pytest',
+            os.path.join(os.path.dirname(__file__), 'convert/meta_data_test_actual.py'), '-s']
+
+    status = subprocess.run(args, env=os.environ, capture_output=True)
+    print("status.stderr")
+    print(status.stderr.decode("utf-8"))
+    print("status.stdout")
+    print(status.stdout.decode("utf-8"))
```
Comment (rkazants) [inline tools/mo/unit_tests/mo/frontend_ngraph_test.py:88]: please remove it. Not sure that it is needed.

File: tools/mo/unit_tests/mo/convert/meta_data_test_actual.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            if key in ignore_attrs:
+                continue
+            assert key in ref_meta, "Unexpected runtime info attribute: {}".format(key)
+
+    def test_meta_data_tf(self):
+        def create_tf_model():
+            #
+            #   Create TF model
+            #
+
```
Comment (rkazants) [inline tools/mo/unit_tests/mo/convert/meta_data_test_actual.py:49]: not needed

File: tools/mo/openvino/tools/mo/convert_impl.py [hunk 3]
```diff
@@ -917,7 +921,10 @@ def _convert(**args):
         init_logger('ERROR', False)
 
         argv.feManager = FrontEndManager()
-        ov_model, legacy_path = driver(argv, {"conversion_parameters": non_default_params})
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/convert_impl.py:920]: please add comments here.
Why can't we check values of parameters during their serialization?

**Review by rkazants — CHANGES_REQUESTED** [review]

File: tools/mo/openvino/tools/mo/convert_impl.py (hunk 3 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/convert_impl.py:920]: I added check here because we can go either legacy path or frontend path. If it is legacy then _tmp IR serializing stucks with model in rt_info, so we need one check there. Another check would be in serialize() method in c++ part. So it looks better to have one check in convert_model.

File: tools/mo/unit_tests/mo/frontend_ngraph_test.py (hunk 1 above)
Comment (popovaan) [reply tools/mo/unit_tests/mo/frontend_ngraph_test.py]: Not needed, removed it.

File: tools/mo/unit_tests/mo/convert/meta_data_test_actual.py (hunk 2 above)
Comment (popovaan) [reply tools/mo/unit_tests/mo/convert/meta_data_test_actual.py]: Removed.

File: tools/mo/openvino/tools/mo/convert_impl.py (hunk 3 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/convert_impl.py]: Yes, this check can be done during serialization. I moved it to rt_info serializing.

File: tools/mo/openvino/tools/mo/convert_impl.py [hunk 4]
```diff
@@ -773,7 +773,11 @@ def pack_params_to_args_namespace(**kwargs):
     for arg in vars(argv):
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/convert_impl.py:773]: Can you please add description for this function? Not cleat what it does?

Comment (rkazants) [inline tools/mo/openvino/tools/mo/convert_impl.py:773]: can you please add more description what `non_default_params` is?

File: tools/mo/unit_tests/mo/convert/meta_data_test_actual.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            ref_meta = ref_meta_data()
+
+            ov_model = convert_model(model)
+            self.check_meta_data(ov_model, ref_meta)
+
+            serialize(ov_model, out_xml.encode('utf-8'), out_xml.replace('.xml', '.bin').encode('utf-8'))
+
+            from openvino.runtime import Core
+            core = Core()
+            serialized_model = core.read_model(out_xml)
```
Comment (rkazants) [inline tools/mo/unit_tests/mo/convert/meta_data_test_actual.py:81]: deserialized_model?

File: tools/mo/unit_tests/mo/convert/meta_data_test_actual.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            self.check_meta_data(serialized_model, ref_meta)
+
+            restored_graph, meta_data = restore_graph_from_ir(out_xml, out_xml.replace('.xml', '.bin'))
+            save_restored_graph(restored_graph, tmpdir, meta_data, "mo_ir_reader_test_model")
+
```
Comment (rkazants) [inline tools/mo/unit_tests/mo/convert/meta_data_test_actual.py:86]: don't we need to check `rt_info` in `restored_graph`?

File: tools/mo/openvino/tools/mo/back/ie_ir_ver_2/emitter.py [hunk 7]
```diff
@@ -427,6 +427,8 @@ def add_net_rt_info(net: Element, meta_info: dict):
     else:
         meta = SubElement(net, 'rt_info')
         for key, value in meta_info.items():
+            if isinstance(value, dict) and value == {}:
+                continue
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/back/ie_ir_ver_2/emitter.py:431]: Not clear why this change is needed? Please clarify

File: tools/mo/openvino/tools/mo/convert_impl.py (hunk 4 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/convert_impl.py:784]: Added comments to pack_params_to_args_namespace() method.

Comment (popovaan) [reply tools/mo/openvino/tools/mo/convert_impl.py:784]: Added comment about `non_default_params`

File: tools/mo/unit_tests/mo/convert/meta_data_test_actual.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
+# Copyright (C) 2018-2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+import tempfile
+import unittest
+from pathlib import Path
+
+from generator import generator
```
[the remark this answers is not in the document]
Comment (popovaan) [reply tools/mo/unit_tests/mo/convert/meta_data_test_actual.py]: No, it is not needed. Removed it.

File: tools/mo/unit_tests/mo/convert/meta_data_test_actual.py (hunk 5 above)
Comment (popovaan) [reply tools/mo/unit_tests/mo/convert/meta_data_test_actual.py]: Renamed it.

File: tools/mo/unit_tests/mo/convert/meta_data_test_actual.py (hunk 6 above)
Comment (popovaan) [reply tools/mo/unit_tests/mo/convert/meta_data_test_actual.py:84]: restored_graph is MO networkx based Graph graph obtained from MO IR reader. It has rt_info, but in another format. 
Here I do the following:
1. read model with MO IR reader 
2. serialize with MO IR reader 
3. read IR to ov.Model and check meta data.

So metadata is checked after obtaining ov.Model from it.

File: tools/mo/openvino/tools/mo/back/ie_ir_ver_2/emitter.py (hunk 7 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/back/ie_ir_ver_2/emitter.py:431]: This check is needed to prevent following line in IR:
```
<rt_info>
<conversion_parameters  />
</rt_info>
```
Such line leads to error of reading of rt_info. 
Before this fix `'input_model`' or `saved_model_dir `was always in conversion_parameters, so such line was impossible. But now '`input_model`' can be skipped and without this check it leads to empty section.

Comment (rkazants) [reply tools/mo/openvino/tools/mo/back/ie_ir_ver_2/emitter.py:431]: Should we report such limitation to IR Frontend?
@ilyachur, fyi.

Comment (popovaan) [reply tools/mo/openvino/tools/mo/back/ie_ir_ver_2/emitter.py:431]: It looks like incorrect format, at least for "conversion_parameters" case, so I don't think it is a bug. Not sure if such empty sections should be supported in general.

File: tools/mo/unit_tests/mo/convert/meta_data_test_actual.py (hunk 6 above)
Comment (rkazants) [reply tools/mo/unit_tests/mo/convert/meta_data_test_actual.py:84]: I understand. We still have POT team work with netwokx based graph. Should we check that this `rt_info` is preserved for them?

File: tools/mo/openvino/tools/mo/convert_impl.py [hunk 9]
```diff
@@ -754,26 +754,41 @@ def args_dict_to_list(cli_parser, **kwargs):
...
     for key, value in kwargs.items():
         if key not in argv and key not in all_params.keys():
             raise Error("Unrecognized argument: {}".format(key))
         if value is not None:
             setattr(argv, key, value)
+
+    # send telemetry with params info
     send_params_info(argv, cli_parser)
 
+    # make dictionary with parameters which have non-default values to be serialized in IR in rt_info
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/convert_impl.py:782]: why should we serialize only parameters with non-default values? Why can't we serialize all parameters applicable for serialization?
I think the logic in this function is overcomplicated: a number of re-packing of parameters.

File: tools/mo/openvino/tools/mo/convert_impl.py [hunk 10]
```diff
@@ -754,26 +754,41 @@ def args_dict_to_list(cli_parser, **kwargs):
...
     fe_manager = FrontEndManager()
     cli_parser = get_all_cli_parser(fe_manager)
+
+    # parse params from convert_model(), the result is argparse.Namespace object
     argv = cli_parser.parse_args(args_dict_to_list(cli_parser, **kwargs))
 
+    # get list of all available params for convert_model()
     all_params = {}
     for key, value in mo_convert_params.items():
         all_params.update(value)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/convert_impl.py:770]: Here I didn't understand how we update the dictionary with just `ParamDescription` object. Can you please clarify?

File: tools/mo/openvino/tools/mo/back/ie_ir_ver_2/emitter.py (hunk 7 above)
Comment (rkazants) [reply tools/mo/openvino/tools/mo/back/ie_ir_ver_2/emitter.py:431]: okay, not principal concern.

File: tools/mo/openvino/tools/mo/convert_impl.py (hunk 9 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/convert_impl.py:782]: We can serialize all parameters as we did before, but there was a request from @slyalin to remove parameters from meta data that are not set by user, because they don't give much information, but make IR look messy.

Regarding this function being overcomplicated, may be split it into sub functions?

File: tools/mo/openvino/tools/mo/convert_impl.py (hunk 10 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/convert_impl.py:770]: There was a change in format of `mo_convert_params` to make it contain groups of parameters like "tf", "kaldi" and so on. To store such information I made `mo_convert_params `a dictionary where key is category ("tf", "onnx"...) and value is dictionary of parameters (key - name, value - `ParamDescription`). 
So in this cycle I get parameters from all categories.

File: tools/mo/unit_tests/mo/convert/meta_data_test_actual.py (hunk 6 above)
Comment (popovaan) [reply tools/mo/unit_tests/mo/convert/meta_data_test_actual.py:84]: As meta data is present after serializing/reading of MO IR reader It means that serializing/reading by MO IR reader does not  affect it. As POT uses MO IR reader for serializing/reading such check looks sufficient.

File: tools/mo/openvino/tools/mo/convert_impl.py (hunk 9 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/convert_impl.py:757]: Refactored this part and removed additional repacking of params. Currently this method returns only non-default params.

File: tools/mo/openvino/tools/mo/convert_impl.py [hunk 11]
```diff
@@ -749,28 +749,18 @@ def args_dict_to_list(cli_parser, **kwargs):
...
-            raise Error("Unrecognized argument: {}".format(key))
-        if value is not None:
-            setattr(argv, key, value)
-    send_params_info(argv, cli_parser)
-
+def get_non_default_params(argv, cli_parser):
+    # make dictionary with parameters which have non-default values to be serialized in IR in rt_info
     non_default_params = {}
-    for arg in vars(argv):
+    for arg in vars(argv).keys():
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/convert_impl.py:755]: can we iterate as follows?

```
for arg, arg_value in vars(argv).items():
```

File: tools/mo/openvino/tools/mo/convert_impl.py [hunk 12]
```diff
@@ -749,28 +749,18 @@ def args_dict_to_list(cli_parser, **kwargs):
...
         arg_value = getattr(argv, arg)
         if arg_value != cli_parser.get_default(arg):
-            non_default_params[arg] = depersonalize(arg_value, arg)
-    return argv, non_default_params
+            value = depersonalize(arg_value, arg)
+            # Skip complex classes in params to prevent
+            # serializing it to rt_info
+            if isinstance(value, (str, bool)):
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/convert_impl.py:761]: should we return other standard types?

File: tools/mo/openvino/tools/mo/convert_impl.py [hunk 13]
```diff
@@ -828,6 +818,8 @@ def show_mo_convert_help():
 
 
 def input_model_is_object(argv):
+    if 'input_model' not in argv:
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/convert_impl.py:821]: may be it is good to leave a comment here that there can be `--saved_model_dir`, or meta specific option which are only used to store paths to the input model.

File: tools/mo/openvino/tools/mo/main.py [hunk 14]
```diff
@@ -66,7 +57,7 @@ def main(cli_parser: argparse.ArgumentParser, framework=None):
             for el in analysis_results.get_messages():
                 log.error(el, extra={'analysis_info': True})
         log.error(err)
-        if not argv['use_new_frontend'] and is_tf:
+        if hasattr(argv, 'use_new_frontend') and not argv.use_new_frontend and is_tf:
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:60]: is it mandatory to check the existence of argument in the dictionary? In the code above you don't check it:) May be `not argv.use_new_frontend and is_tf` is enough?

File: tools/mo/openvino/tools/mo/convert_impl.py [hunk 15]
```diff
@@ -881,19 +873,46 @@ def _convert(**args):
...
+
+                # Non string params like input_model or extensions are ignored by parse_args()
+                # so we need to set them in argv separately
+                if value is not None and getattr(argv, key) != value:
+                    setattr(argv, key, value)
+        else:
+            argv = cli_parser.parse_args()
+
+        if framework is not None:
+            setattr(argv, 'framework', framework)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/convert_impl.py:906]: I think we need to move this into separate routine like `pack_params_to_args_namespace` and test it well.

File: tools/mo/openvino/tools/mo/main.py (hunk 14 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/main.py:60]: In code above I don't check it because we have lines before try-catch block
`argv = cli_parser.parse_args()`
...
`argv = vars(argv)`
`argv = cli_parser.parse_args()` returns namespace with all mo params even if they are not set. Not set params have `None `values.
`argv = vars(argv)` converts Namespace to dictionary which has all parameters as well. 

But in new code we have  `argv = None`  before try-catch block.
If convert fails before returning parsed dictionary with params then argv remains with None. So in this case this check will cause error.

File: tools/mo/openvino/tools/mo/convert_impl.py (hunk 11 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/convert_impl.py]: Yes, done.

File: tools/mo/openvino/tools/mo/convert_impl.py (hunk 12 above)
Comment (popovaan) [reply tools/mo/openvino/tools/mo/convert_impl.py]: Most of the parameters are converted to string before this line, but for 'batch' and 'scale' conversion to string is not happening, they keep numeric values. So I added numbers.Number to this check and added 'batch' and 'scale' to meta data test.

File: tools/mo/openvino/tools/mo/main.py [hunk 16]
```diff
@@ -32,29 +32,20 @@ def main(cli_parser: argparse.ArgumentParser, framework=None):
...
-    if framework is not None:
-        argv['framework'] = framework
 
     ngraph_function = None
+    argv = None
+    is_tf = False
     try:
-        ngraph_function = convert_model(**argv)
+        ngraph_function, argv = _convert(cli_parser, framework, {})
+        is_tf, _, _, _, _ = deduce_legacy_frontend_by_namespace(argv)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:41]: please swap these two lines because in case of failure of conversion of TF model using TF FE `is_tf` is False so we cannot raise info message. Please check it.

Comment (popovaan) [reply tools/mo/openvino/tools/mo/main.py:41]: But we don't have argv until we run `_convert()`. To obtain args from cli_parser we need to run `parse_args()` which is moved to `_convert()`.

Comment (popovaan) [reply tools/mo/openvino/tools/mo/main.py:41]: By the way, do we still need to print get_tf_fe_legacy_message() if we plan to switch to TF FE by default?

Comment (rkazants) [reply tools/mo/openvino/tools/mo/main.py:41]: ok, agreed. I will remove it in my PR for switching to TF FE.

**Review by rkazants — APPROVED** [review]
